### ISSUE TITLE
`amp-date-display` inside `amp-list` does not render correctly

### ISSUE THREAD

**Issue by dritter** [ampproject/amphtml#23731]
There are multiple problems with `amp-date-display` inside an `amp-list`:

1. The validator fails to detect `amp-date-display` in mustache templates ("The extension 'amp-date-display' was found on this page, but is unused. Please remove this extension."):
```html
<!doctype html>
<html ⚡>
<head>
  <meta charset="utf-8">
  <title>My AMP Page</title>
  <link rel="canonical" href="self.html" />
  <meta name="viewport" content="width=device-width,minimum-scale=1">
  <style amp-boilerplate>body{-webkit-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-moz-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-ms-animation:-amp-start 8s steps(1,end) 0s 1 normal both;animation:-amp-start 8s steps(1,end) 0s 1 normal both}@-webkit-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-moz-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-ms-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-o-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}</style><noscript><style amp-boilerplate>body{-webkit-animation:none;-moz-animation:none;-ms-animation:none;animation:none}</style></noscript>
  <script async src="https://cdn.ampproject.org/v0.js"></script>
  <script async custom-element="amp-list" src="https://cdn.ampproject.org/v0/amp-list-0.1.js"></script>
  <script async custom-template="amp-mustache" src="https://cdn.ampproject.org/v0/amp-mustache-0.2.js"></script>
  <script async custom-element="amp-date-display" src="https://cdn.ampproject.org/v0/amp-date-display-0.1.js"></script>
  <style amp-custom>
    h1 {
      margin: 1rem;
    }
  </style>
</head>
<body>
  <h1>Hello AMPHTML World!</h1>
  <amp-list src="https://cors-test.appspot.com/test"
            single-item
            items="."
            layout="fixed-height"
            height="50">
    
    <script type="text/plain" template="amp-mustache">
      <amp-date-display datetime="2006-02-14T15:27:33+01:00" layout="fixed-height" height="20">
      	<template type="amp-mustache">  	
          <small>{{ day }}.{{ month }}.{{ year }} {{ hourTwoDigit }}:{{ minuteTwoDigit }}</small>
		</template>
      </amp-date-display>
    </script>
  </amp-list>
  
</body>
</html>
```

2. `amp-date-display` fails to render if written inside of an `amp-list`:
![Screenshot_20190806_172540](https://user-images.githubusercontent.com/1544760/62553265-4f21b300-b86f-11e9-94df-63cecf81187a.png)

```html
<!doctype html>
<html ⚡>
<head>
  <meta charset="utf-8">
  <title>My AMP Page</title>
  <link rel="canonical" href="self.html" />
  <meta name="viewport" content="width=device-width,minimum-scale=1">
  <style amp-boilerplate>body{-webkit-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-moz-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-ms-animation:-amp-start 8s steps(1,end) 0s 1 normal both;animation:-amp-start 8s steps(1,end) 0s 1 normal both}@-webkit-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-moz-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-ms-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-o-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}</style><noscript><style amp-boilerplate>body{-webkit-animation:none;-moz-animation:none;-ms-animation:none;animation:none}</style></noscript>
  <script async src="https://cdn.ampproject.org/v0.js"></script>
  <script async custom-element="amp-list" src="https://cdn.ampproject.org/v0/amp-list-0.1.js"></script>
  <script async custom-template="amp-mustache" src="https://cdn.ampproject.org/v0/amp-mustache-0.2.js"></script>
  <script async custom-element="amp-date-display" src="https://cdn.ampproject.org/v0/amp-date-display-0.1.js"></script>
  <style amp-custom>
    h1 {
      margin: 1rem;
    }
  </style>
</head>
<body>
  <h1>Hello AMPHTML World!</h1>
  <amp-list src="https://cors-test.appspot.com/test"
            single-item
            items="."
            layout="fixed-height"
            height="50">
    
    <script type="text/plain" template="amp-mustache">
      This should be a proper formatted date:
      <amp-date-display datetime="2006-02-14T15:27:33+01:00" layout="fixed-height" height="20">
      	<template type="amp-mustache">  	
          <small>{{ day }}.{{ month }}.{{ year }} {{ hourTwoDigit }}:{{ minuteTwoDigit }}</small>
		</template>
      </amp-date-display>
    </script>
  </amp-list>
  This is outside of amp-list:
  <amp-date-display datetime="2012-02-14T15:27:33+01:00" layout="fixed-height" height="20">
    <template type="amp-mustache">  	
      <small>{{ day }}.{{ month }}.{{ year }} {{ hourTwoDigit }}:{{ minuteTwoDigit }}</small>
    </template>
  </amp-date-display>
</body>
</html>
```

3. `amp-date-display` fails to render completely if the template is written in the alternative syntax (`<script type="text/plain" template="amp-mustache">`). This is not allowed by the validator and probably a general error. <script> inside <template> is not allowed. Only the other way around is allowed.
```html
<!doctype html>
<html ⚡>
<head>
  <meta charset="utf-8">
  <title>My AMP Page</title>
  <link rel="canonical" href="self.html" />
  <meta name="viewport" content="width=device-width,minimum-scale=1">
  <style amp-boilerplate>body{-webkit-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-moz-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-ms-animation:-amp-start 8s steps(1,end) 0s 1 normal both;animation:-amp-start 8s steps(1,end) 0s 1 normal both}@-webkit-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-moz-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-ms-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-o-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}</style><noscript><style amp-boilerplate>body{-webkit-animation:none;-moz-animation:none;-ms-animation:none;animation:none}</style></noscript>
  <script async src="https://cdn.ampproject.org/v0.js"></script>
  <script async custom-element="amp-list" src="https://cdn.ampproject.org/v0/amp-list-0.1.js"></script>
  <script async custom-template="amp-mustache" src="https://cdn.ampproject.org/v0/amp-mustache-0.2.js"></script>
  <script async custom-element="amp-date-display" src="https://cdn.ampproject.org/v0/amp-date-display-0.1.js"></script>
  <style amp-custom>
    h1 {
      margin: 1rem;
    }
  </style>
</head>
<body>
  <h1>Hello AMPHTML World!</h1>
  <amp-list src="https://cors-test.appspot.com/test"
            single-item
            items="."
            layout="fixed-height"
            height="50">
    <template type="amp-mustache">
      <amp-date-display datetime="2006-02-14T15:27:33+01:00" layout="fixed-height" height="20">
      	<script type="text/plain" template="amp-mustache">
        
          <small>{{ day }}.{{ month }}.{{ year }} {{ hourTwoDigit }}:{{ minuteTwoDigit }}</small>
        </script>
      </amp-date-display>
    </template>
  </amp-list>
  
</body>
</html>
```

**Comment by cathyxz** [comment]
I think this is because you cannot have nested templates in `<amp-mustache>`. See: https://amp.dev/documentation/components/amp-mustache/#nested-templates. 

You can workaround this by referencing a template via its `id`. 

For example, altering your above code to: 

```
<!doctype html>
<html ⚡>
<head>
  <meta charset="utf-8">
  <title>My AMP Page</title>
  <link rel="canonical" href="self.html" />
  <meta name="viewport" content="width=device-width,minimum-scale=1">
  <style amp-boilerplate>body{-webkit-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-moz-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-ms-animation:-amp-start 8s steps(1,end) 0s 1 normal both;animation:-amp-start 8s steps(1,end) 0s 1 normal both}@-webkit-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-moz-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-ms-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-o-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}</style><noscript><style amp-boilerplate>body{-webkit-animation:none;-moz-animation:none;-ms-animation:none;animation:none}</style></noscript>
  <script async src="https://cdn.ampproject.org/v0.js"></script>
  <script async custom-element="amp-list" src="https://cdn.ampproject.org/v0/amp-list-0.1.js"></script>
  <script async custom-template="amp-mustache" src="https://cdn.ampproject.org/v0/amp-mustache-0.2.js"></script>
  <script async custom-element="amp-date-display" src="https://cdn.ampproject.org/v0/amp-date-display-0.1.js"></script>
  <style amp-custom>
    h1 {
      margin: 1rem;
    }
  </style>
</head>
<body>
  <h1>Hello AMPHTML World!</h1>
  <amp-list src="https://cors-test.appspot.com/test"
            single-item
            items="."
            layout="fixed-height"
            height="50">

    <script type="text/plain" template="amp-mustache">
      This should be a proper formatted date:
      <amp-date-display datetime="2006-02-14T15:27:33+01:00" layout="fixed-height" height="20" template="date-template"> </amp-date-display>
    </script>
  </amp-list>

  <template type="amp-mustache" id="date-template">
    <small>{{ day }}.{{ month }}.{{ year }} {{ hourTwoDigit }}:{{ minuteTwoDigit }}</small>
</template>
  
</body>
</html>
```

Live codepen: https://codepen.io/cathyxz/pen/vodZZQ?editors=1000

**Comment by dritter** [comment]
@cathyxz Thanks. This works, but _only_ if the outer template is written as `<script type="text/plain"`, otherwise the validator shows:

> The attribute 'template' may not appear in tag 'amp-date-display'. (see https://amp.dev/documentation/components/amp-date-display)

**Comment by cathyxz** [comment]
Okay, given that workaround, I think there are a few separate bugs here: 

1. `<amp-date-display>` does not render correctly if you put the reference `<template>` in the page below the `<amp-list>`. 

### Does NOT Render

```
  <amp-list src="https://cors-test.appspot.com/test"
            single-item
            items="."
            layout="fixed-height"
            height="50">

    <template type="amp-mustache">
      This should be a proper formatted date:
      <amp-date-display datetime="2006-02-14T15:27:33+01:00"
                        layout="fixed-height"
                        height="20"
                        template="date-template">
      </amp-date-display>
    <template/>
  </amp-list>

  <template type="amp-mustache" id="date-template">
      <small>{{ day }}.{{ month }}.{{ year }} {{ hourTwoDigit }}:{{ minuteTwoDigit }}</small>
  </template>

```

<img width="418" alt="Screenshot 2019-08-07 at 4 01 26 PM" src="https://user-images.githubusercontent.com/1528181/62663529-b2bde600-b92c-11e9-8781-9889ef8b3df9.png">


### Does Render
```
  <amp-list src="https://cors-test.appspot.com/test"
            single-item
            items="."
            layout="fixed-height"
            height="50">

    <template type="amp-mustache">
      This should be a proper formatted date:
      <amp-date-display datetime="2006-02-14T15:27:33+01:00"
                        layout="fixed-height"
                        height="20"
                        template="date-template">
      </amp-date-display>
    <template/>
  </amp-list>

  <template type="amp-mustache" id="date-template">
      <small>{{ day }}.{{ month }}.{{ year }} {{ hourTwoDigit }}:{{ minuteTwoDigit }}</small>
  </template>
```

2. We need to allow the `template` attribute on `<amp-date-display>` in the validator. This is a simple fix. 

3. Possibly need to allow `template` attribute as a descendent of `<template>` tag, or to clearly document if the correct use case for this is to use the `<script>` tag version of `amp-mustache`.

**Comment by dritter** [comment]
#### Error 1
I can hardly see any difference between your examples.. I assume in the latter one the `<template>` Tag should be replaced with the `<script>` tag version?

Btw there is a mistake in the closing `</template>` tag (`<template/>`).

#### Error 2
Agreed.

#### Error 3
Agreed x2

Additionally the validator should discover `amp-date-display` inside of the `script` version of the template tag (my very first example in this issue):
![](https://user-images.githubusercontent.com/1544760/62664467-e542ff80-b97b-11e9-825c-c2d61e391c7c.png)

**Comment by dritter** [comment]
Thanks @cathyxz
